aten/src/ATen/native/mps/operations/Attention.mm

```diff
@@ -0,0 +1,111 @@
+#include <string>
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <iostream>
+#include <optional>
+
+#include <ATen/core/Tensor.h>
+#include <ATen/native/mps/OperationUtils.h>
+#include <ATen/native/transformers/attention.h>
+#include <ATen/native/transformers/sdp_utils_cpp.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_scaled_dot_product_attention_math_for_mps_native.h>
+#include <ATen/ops/empty_native.h>
+#endif
+
+namespace at::native {
+
+std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math_mps(const Tensor& query,
+                                                                  const Tensor& key,
+                                                                  const Tensor& value,
+                                                                  const std::optional<Tensor>& attn_mask,
+                                                                  double dropout_p,
+                                                                  bool is_causal,
+                                                                  const std::optional<Tensor>& dropout_mask,
+                                                                  std::optional<double> scale) {
+  using namespace mps;
+  struct CachedGraph : public MPSCachedGraph {
+    CachedGraph(MPSGraph* graph) : MPSCachedGraph(graph) {}
+    MPSGraphTensor* qTensor = nil;
+    MPSGraphTensor* kTensor = nil;
+    MPSGraphTensor* vTensor = nil;
+    MPSGraphTensor* maskTensor = nil;
+    MPSGraphTensor* outputTensor = nil;
+    MPSGraphTensor* attnTensor = nil;
+  };
+  int64_t batchSize = query.size(0);
+  int64_t num_head = query.size(1);
+  int64_t qSize = query.size(2);
+  int64_t headSize = query.size(3);
+  int64_t maxSeqLength = key.size(2);
+  auto out = at::empty({batchSize, num_head, qSize, headSize}, query.options());
+  auto attn = at::empty({batchSize, num_head, qSize, maxSeqLength}, query.options());
+  auto scale_factor = sdp::calculate_scale(query, scale).sqrt().as_float_unchecked();
+  @autoreleasepool {
+    auto mkey = __func__ + getTensorsStringKey({query, key, value}) + ":" + std::to_string(is_causal) + ":" + std::to_string(attn_mask.has_value());
+    auto cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>(mkey, [&](auto mpsGraph, auto graph) {
+      auto qTensor = mpsGraphRankedPlaceHolder(mpsGraph, query);
+      auto kTensor = mpsGraphRankedPlaceHolder(mpsGraph, key);
+      auto vTensor = mpsGraphRankedPlaceHolder(mpsGraph, value);
+      auto kT = [mpsGraph transposeTensor:kTensor dimension:2 withDimension:3 name:nil];
+      auto scaleTensor = [mpsGraph constantWithScalar:scale_factor shape:getMPSShape({1}) dataType:MPSDataTypeFloat32];
+
+      // sqTensor = qTensor * scale_factor
+      auto castedQ = qTensor;
+      if ([qTensor dataType] != MPSDataTypeFloat32) {
```

Comment:
Also do you wanna assert on supported dtypes being fp16 and fp32 only? or bf16 is also supported?

Reply:
bf16 is also supported